### ISSUE TITLE
Number (Repeat X Times ) + Arrow Keys not always handled correctly

### ISSUE THREAD

**Issue by tgrelka** [gwsw/less#116]
I’m using `less` Version 563 installed using the scoop package manager on Windows 10 20H2 (Build 19042.746)

While using arrow keys from the neo2 layout on Mod4 an entering a number to move X amount of lines, instead of moving e.g. 10 times (10<Mod4+a> would be what I typed), I move just one line. All other arrow keys work fine, but for some reason `less` does not repeat the action 10 times.
It should be of note that I’m using the kbdneo2 driver with the AHK supplement, the latter is responsible for the arrow key inputs actually send out (using AHK built-in 'up', 'down' etc). It seems like `less` does partially recognize the inputs as up, down, (left, right) as it is supposed to but somehow does not handle them completely because it discards any numbers typed before them. That’s why I suspect it isn’t necessarily a problem with AHK or anything.

I might have some free time later to investigate this issue, but I am not familiar with the codebase so I can’t promise anything.

Cheers! 🎊

**Comment by gwsw** [comment]
That's strange. Less doesn't have any provision for a movement command that doesn't accept a number argument, except for mouse scrolling. If you can, I would first confirm that your arrow key is causing less to reach the case A_F_LINE at line 1293 in command.c, and that the value of the "number" variable is the number you entered.

**Comment by tgrelka** [comment]
Yeah, I also found it really strange that the movement is recognized, but not the number.
I’ll set up a build environment later and try to debug the whole thing, thank you for your insight! I might come back if I need any help, I rarely code in C on my Windows machine and as such have no toolchain set up.

**Comment by tgrelka** [comment]
So after a bit of fiddling I managed to successfully compile less 563 using the new Makefile.wng and mingw64 (installed via msys2).
Debugging with gdb, I set a breakpoint at 1293 (automatically moved to 1297 `if (number <= 0)`) and as it turns out, `number` is 0 at that point, even though I’ve seen it reach 10.
If I type (in less) `10<Enter>`, I reach that same point, but with number being 10.

**Comment by tgrelka** [comment]
![image](https://user-images.githubusercontent.com/22134144/105155765-44aeb400-5b0b-11eb-8aba-fec9c579a28d.png)
This is my output of gdb, with a simple script at the breakpoint to print value of number and then continue.

**Comment by gwsw** [comment]
What are the actual characters sent by your arrow keys in your environment? Perhaps less is misinterpreting part of the arrow sequence as a command to clear the line number. If you're using terminfo, you can use these commands to print the up and down arrow key sequences:
```
tput kcuu1 | od -c
tput kcud1 | od -c
```

**Comment by tgrelka** [comment]
These are sent out by AutoHotKey (AHK) die to the neo2 layout driver not (yet) able to send these natively.
I'd have to look at the AHK source code, but I've checked the script, which sends nothing but the built-in "up", "down" etc, so whatever keycodes AHK sends there seem to be the problem - surprisingly they are in no other software. I use (neo)vim on a daily basis and they work just fine there, as well as every software I've configured to use vi(m)-like keybindings.

So while I don't fully blame less for the issue, I think there's still something unintended going on.

I'll take a look at the keycodes and also at what AHK sends out.

**Comment by tgrelka** [comment]
By all other arrow keys I meant the native arrow keys between numpad and the default keys, as well as the arrow keys on the numpad itself.

**Comment by tgrelka** [comment]
> What are the actual characters sent by your arrow keys in your environment? Perhaps less is misinterpreting part of the arrow sequence as a command to clear the line number. If you're using terminfo, you can use these commands to print the up and down arrow key sequences:
> 
> ```
> tput kcuu1 | od -c
> tput kcud1 | od -c
> ```
I’m using Powershell Core 7.1, the commands give the following output:
```
0000000 033   O   A  \r  \n
0000005

0000000 033   O   B  \r  \n
0000005
```

**Comment by gwsw** [comment]
That's strange; those are pretty standard sequences for arrow keys. If I'm understanding your description correctly, it's possible that AHK is sending something different than these sequences, is that correct? I guess if I had a way to reproduce this, my next step would be to run under a debugger, set a breakpoint in getcc() and see the sequence received when you press the arrow key. If that doesn't immediately reveal the issue, I would then trace the flow through the commands() function and see where "number" is getting cleared (most likely at the top of the for loop, but it should get there only after processing all digits of the number plus the following command). I don't know how much of this is feasible for you.

**Comment by tgrelka** [comment]
@gwsw I’ve debugged less before (see my comment from two weeks ago), I don’t think it’s much trouble doing so again. I’ll do that later - I’ve looked at the AHK code base already but couldn’t really find anything in their code directly, it seems like an abstraction layer of Windows is responsible for sending the actual sequence. Which is weird, because basically that’s tasked to send e.g. an „arrow up“ sequence.
Once I’ve got less running with a debugger again, I’ll look at getcc() and let you know of the results!
I don’t really use C apart from some hobby projects with micro controllers, but I’m familiar enough with many other languages that I think I’ll figure it out.

**Comment by tgrelka** [comment]
Hm, I actually do need some help - where do I best set breakpoints and which values should I inspect to get the char sequence?

**Comment by gwsw** [comment]
I think the easiest approach would be to set the breakpoint in getcc() and look at the returned value. You may want to temporarily edit the code to store the value in a variable before returning it, to make it easier to inspect.

**Comment by tgrelka** [comment]
Well, weirdly enough I can no longer reproduce this issue.
The only thing having been updated between then and now is Windows, so I’m guessing it’s been a problem with Windows sending weird codes when tasked to send an arrow down.

I haven’t changed anything else since yesterday, but while debugging today everything worked fine. I thought maybe I had encountered [heisenbug](https://en.wikipedia.org/wiki/Heisenbug), but even after removing all traces of the less source and using the official binary, it worked fine everywhere.

My procedure to debug the getcc() function in gdb was simply `b getcc`, setting the breakpoint, `commands`, then use `finish`, `cont` and `end`, then testing in a separate less instance with gdb attached. (Mainly for future reference - I usually don’t use gdb.)

Regarding building with debug symbols, I’ll open another issue.
